contracts/Validator.sol

```diff
@@ -0,0 +1,301 @@
+    /// @dev  Stake allows to stake on the Polygon stakeManager contract by
+    /// calling stakeFor function and set the user address equal to this contract address
+    /// @param _sender the address of the operator-owner that approved Matics.
+    /// @param _amount the amount to stake with.
+    /// @param _heimdallFee the heimdall fees.
+    /// @param _acceptDelegation accept delegation.
+    /// @param _signerPubkey signer public key used on the heimdall node.
+    /// @param _commissionRate validator commision rate
+    /// @return Returns the validatorId and the validatorShare contract address.
+    function stake(
+        address _sender,
+        uint256 _amount,
+        uint256 _heimdallFee,
+        bool _acceptDelegation,
+        bytes memory _signerPubkey,
+        uint256 _commissionRate
+    ) external override isOperator returns (uint256, address) {
+        // get operator
+        INodeOperatorRegistry operator = getOperator();
+
+        // get stakeManager
+        IStakeManager stakeManager = IStakeManager(operator.getStakeManager());
+
+        uint256 totalAmount = _amount + _heimdallFee;
+        // approve Polygon token to stake manager totalAmount
+        IERC20 polygonERC20 = IERC20(operator.getPolygonERC20());
+
+        // transfer tokens from user to this contract
+        polygonERC20.safeTransferFrom(
+            _sender,
+            address(this),
+            (_amount + _heimdallFee)
+        );
+
+        // approve to stakeManager
+        polygonERC20.safeApprove(address(stakeManager), totalAmount);
+
+        // call polygon stake manager
+        stakeManager.stakeFor(
+            address(this),
+            _amount,
+            _heimdallFee,
+            _acceptDelegation,
+            _signerPubkey
+        );
+        // get validator id
+        uint256 validatorId = stakeManager.getValidatorId(address(this));
+        // get validatorShare contract
+        address validatorShare = stakeManager.getValidatorContract(validatorId);
+        // set commision
+        stakeManager.updateCommissionRate(validatorId, _commissionRate);
+
+        return (validatorId, validatorShare);
+    }
+
+    /// @notice Restake Matics for a validator on polygon stake manager.
+    /// @param _sender operator owner which approved tokens to the validato contract.
+    /// @param _validatorId validator id.
+    /// @param _amount amount to stake.
+    /// @param _stakeRewards restake rewards.
+    function restake(
+        address _sender,
+        uint256 _validatorId,
+        uint256 _amount,
+        bool _stakeRewards
+    ) external override isOperator {
+        // get operator
+        INodeOperatorRegistry operator = getOperator();
+
+        // get stakeManager
+        address stakeManager = operator.getStakeManager();
+
+        // approve Polygon token to stake manager totalAmount
+        IERC20 polygonERC20 = IERC20(operator.getPolygonERC20());
```

Comment:
Yes, we could but for testing, it will be difficult because you need to have the right token. Also, you can see that I regularly do calls to the operator the reason is I tried to maintain the state(all the addresses like erc20 or stakeManager...) on the operator to avoid duplicates..